### PR TITLE
AP_NavEKF3: Accumulate error for cores

### PR DESCRIPTION
**Pull request by harshitsankhla**
To improve core selection logic, accumulate error for each EKF lane then select the one with lowest error as new primary.

We start by simply adding up NavEKF3_core::errorScore() for each lane. We would like to explore a better metric that takes into account - 
1. State Difference
2. Innovations
3. Variances

Finally, the error should be some weighted combination of factors can quantify a core's error.

### CODE REVIEW

File: libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp [hunk 1]
```diff
@@ -51,6 +51,16 @@ float NavEKF3_core::errorScore() const
     return score;
 }
 
+void NavEKF3_core::updateCoreError()
+{
+    runningError += NavEKF3_core::errorScore();
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:56]: this looks like runningError will always increase. I think we need a number that is on the same scale throughout the flight, so we can have meaningful thresholds.
If you leave the log up in AP_NavEKF3.cpp then I think you could accumulate the difference between the error score of each lane and the error score of the primary. That number would always be zero for the primary, and would grow as +ve or -ve number for lanes that are worse or better than the primary.

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 2]
```diff
@@ -815,23 +815,29 @@ void NavEKF3::UpdateFilter(void)
...
         runCoreSelection = (imuSampleTime_us - lastUnhealthyTime_us) > 1E7;
     }
+
+    //start updating errors for all cores after core selection logic is true
+    if(runCoreSelection) {
+        updateCoreErrors();
+    }
+
     float primaryErrorScore = core[primary].errorScore();
     if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:825]: this if statement means we only consider lane changes if the primary is unhealthy or it's score is high. I think we need to consider switching if the accumulated difference between the primary and an alternative lane goes over a threshold

File: libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp (hunk 1 above)
Comment (harshitsankhla) [reply libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:56]: Yes, the unbounded error is an issue, I was thinking about some normalisation to constrain it in a range but could not figure out logically. I'll modify this to accumulating differences with respect to the primary. Also when we switch lanes, should I reset all the errors by making the new primary's error 0 or continue with the existing, which also is fine as error would still accumulate in the respective lanes, just there would be an initial offset which I don't know if it would create any problems.

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 2 above)
Comment (harshitsankhla) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp:825]: I think that logic is this one just below it - 
`if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection)`
what the new if condition stands for is to start accumulating errors only after core selection logic is available (the if block just above it does that). The reason for waiting till core selection is available is to not accumulate errors that happen due to the initialisation process. Please correct me if I'm wrong. 
I think we should add a separate check that ensures even when the current primary is healthy or is returning an NavEKF_core::errorScore() < 1.0f,  if difference between accumulated error between primary and another lane is high above a threshold, we should switch.

File: libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:56]: I think reset all to zero on switch

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp:825]: sounds good, thanks

**Comment by harshitsankhla** [comment]
Status as on June 16, 2020 -

1. Introduced error accumulation for all active cores, relative to the primary core at the instance. Reset core errors to zero when we switch.
2. Reworked the core selection to logic to first search for the best core and then switch to it if criteria is met.
3. Added logging for relative core errors to XKF5. Created a separate log structure to avoid conflict with EKF2's NKF5.
4. Tested with 3 IMUs in SITL.

Some results from @tridge -

Screenshot from 2020-06-16 14-41-33
Screenshot from 2020-06-16 14-41-55

Some timing estimates for different thresholds - https://docs.google.com/spreadsheets/d/1UGZopVCcp5afHvm16RUu-mL7smuukvso__V2CIirCl4/edit?usp=sharing

File: libraries/AP_Logger/LogStructure.h [hunk 3]
```diff
@@ -571,6 +571,24 @@ struct PACKED log_NKF5 {
...
+    int16_t AFI;
+    int16_t HAGL;
+    int16_t offset;
+    int16_t RI;
+    uint16_t meaRng;
+    uint16_t errHAGL;
+    float angErr;
+    float velErr;
+    float posErr;
+    float RErr[3]; // set according to defined maximum allowed EKF3 cores
```
Comment (tridge) [inline libraries/AP_Logger/LogStructure.h:589]: Intead of the new structure I think this should go into XKF3. Sorry for suggesting XKF5 on discord. We can do it as a single float on XKF3 as that msg is per-lane. 
Note that while the XKF3 structure is shared with EKF2, I think that is OK, as we likely will bring across your new lane switching code to EKF2 as well

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #include <AP_GPS/AP_GPS.h>
 #include <AP_VisualOdom/AP_VisualOdom.h>
 #include <new>
+// #include <stdio.h>
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:10]: should remove this

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 5]
```diff
@@ -815,34 +822,53 @@ void NavEKF3::UpdateFilter(void)
...
         runCoreSelection = (imuSampleTime_us - lastUnhealthyTime_us) > 1E7;
     }
-    float primaryErrorScore = core[primary].errorScore();
-    if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection) {
-        float lowestErrorScore = 0.67f * primaryErrorScore;
-        uint8_t newPrimaryIndex = primary; // index for new primary
-        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
+    
+                                   
+    if(runCoreSelection) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:827]: should have space after if in ArduPilot style

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 6]
```diff
@@ -815,34 +822,53 @@ void NavEKF3::UpdateFilter(void)
...
-                // If the primary core is still healthy,then switching is optional and will only be done if
-                // a core with a significantly lower error score can be found
-                float altErrorScore = core[coreIndex].errorScore();
-                if (altCoreAvailable && (!core[newPrimaryIndex].healthy() || altErrorScore < lowestErrorScore)) {
+                float altCoreError = relativeCoreError[coreIndex];
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:846]: I think we should consider dropping the statePredictEnabled check. That check made sense when we were making a decision based on a single sample, but makes less sense with the accumulated error. Paul, what do you think?

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 7]
```diff
@@ -907,6 +933,35 @@ void NavEKF3::requestYawReset(void)
...
+/*
+  Update the error for all running cores. Error is accumulated relative to the primary core's error.
+  A positive relative error for a core means it has been more erroneous than the existing primary.
+  A negative relative error indicates a better core which can be switched to.
+*/
+void NavEKF3::updateCoreErrors(float primaryErrorScore)
+{
+    float error = 0;
+    for (uint8_t i = 0; i < num_cores; i++) {
+        if(i!=primary) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:945]: whitespace

File: libraries/AP_Logger/LogStructure.h (hunk 3 above)
Comment (harshitsankhla) [reply libraries/AP_Logger/LogStructure.h]: Changed. In the meantime for EKF2, since we are using the same NKF3 structure, setting the error value to be 0. Hope that’s fine for now.

**Comment by tridge** [comment]
getting close, thanks. We do want one commit per folder, so just run Tools/gittools/git-subsystems-split to split it

**Comment by tridge** [comment]
we're also going to need some flight tests of this on a real vehicle. I could fly this on a small quad

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 8]
```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
...
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
+
+                if(altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a 
+                    // better core and would like to switch to it even if the current primary is healthy
+                    betterCore = (0.5*CORE_ERR_LIM + altCoreError) < relativeCoreError[primary]; // consider a better core if its relative error is lower by half the max possible error
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:850]: I think the 0.5*CORE_ERR_LIM should be replaced by a separate threshold define. The CORE_ERR_LIM and this threshold are not really related.

**Comment by tridge** [comment]
I've been testing this in SITL and I'm not seeing any case where we get the gradual buildup of errors then a switch. I don't know if this is just the thresholds, or if there is something more fundamental.
Note that the EKF logging for plane is only 25Hz. Changing it to be full rate by changing the Log_Write_Fast rate in Plane::scheduler_tasks[] makes it a bit clearer.
This is what I see with a X gyro offset error of 0.02:
image
that is with SCHED_LOOP_RATE=400, so each data point is 2.5ms.
the lane switch happened 2s after the parameter change, but we only see the RErr start to build up 12ms before the switch. 
log here:
http://uav.tridgell.net/EKF/00000005.BIN

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 9]
```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
...
-        // update the yaw and position reset data to capture changes due to the lane switch
-        if (newPrimaryIndex != primary) {
+        altCoreAvailable = newPrimaryIndex != primary;
+
+        // Switch cores if another core is available and the active primary core meets one of the following conditions - 
+        // 1. has a bad error score
+        // 2. is unhealthy
+        // 3. is healthy, but a better core is available
+        // also update the yaw and position reset data to capture changes due to the lane switch
+        if (altCoreAvailable && (primaryErrorScore > 1.0f || !core[primary].healthy() || betterCore)) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:863]: we've lost the send_text() for the lane switch which we have in master

**Comment by tridge** [comment]
I think it would be helpful to also log errorScore(), so we can see if the relative error is performing as expected

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 9 above)
Comment (harshitsankhla) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp:863]: master doesn't seem to have a send_text() in UpdateFilter(), its in checkLaneSwitch() here - https://github.com/ArduPilot/ardupilot/blob/ca0ae57b56f472ef656ae82eb0a8afe42217c2de/libraries/AP_NavEKF3/AP_NavEKF3.cpp#L899
I'll copy it over. Does this mean that the Lane Switch messages we were seeing before were not because of switches that happened in UpdateFilter() ? I think we might to rework checkLaneSwitch() as well for the new relative error.

**Comment by harshitsankhla** [comment]
Resolved the above !

Some thoughts, should we - 
1. consider having a minimum time required between 2 consecutive switches ?
2. checkLaneSwitch() also be reworked to use the relative error terms ?
3. send_text() if no better lane is found ?

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 6 above)
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp:845]: In practice it should make little difference now given real world noise and sensor errors, but in theory if there is a sensor error transient that effects both lanes, removing it will disadvantage the leading lane. It's an easy change to test so worth a try.

**Comment by tridge** [comment]
failing doc tests

**Comment by harshitsankhla** [comment]
> failing doc tests

fixed, sorry

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 10]
```diff
@@ -815,34 +821,54 @@ void NavEKF3::UpdateFilter(void)
...
-                if (altCoreAvailable && (!core[newPrimaryIndex].healthy() || altErrorScore < lowestErrorScore)) {
+                float altCoreError = coreRelativeErrors[coreIndex];
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
+
+                if (altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a
```
Comment (priseborough) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:848]: This will need some follow-on replay tuning to avoid unnecessary resets. Switching does produce some controller transients so we don't want to be doing them frequently.

**Comment by priseborough** [comment]
Can we get a branch withe this PR and https://github.com/ArduPilot/ardupilot/pull/14674 so they can be assessed together? I'm not confident I understand how the combination of the two will work.

**Comment by harshitsankhla** [comment]
> Can we get a branch with this PR and #14674 so they can be assessed together? I'm not confident I understand how the combination of the two will work.

@priseborough #14674 has this PR's accumulated error logic included. The branch is - https://github.com/harshitsankhla/ardupilot/tree/ekf3-affinity

**Comment by harshitsankhla** [comment]
@priseborough 

This is my understanding of the Affinity and Accumulated Error Lane Switching. Also leaving this here for the general audience.

Affinity - For a vehicle which has multiple (Airspeed - Magnetometer - Barometer - GPS) affinity is a way to assign these sensors between separate EKF cores. If a sensor is healthy and can be used for fusion, we want it to be attached to a specific core. If the number of cores is more than the number of sensor instances, then some cores may share the same sensor (specifically the primary one). Considering a vehicle which has 2 baros , 2 GPS, 2 airspeeds, 3 mags and 3 IMUs - following would be the allotment if affinity is enabled and all sensors are healthy.

LANE  |   1  |   2  |  3 
-----------------------------
ASD    |   1  |   2  |   1
BAR    |   1  |   2  |   1
GPS    |   1  |   2  |   1
MAG   |   1  |   2  |   3

Accumulated Error - The error which we accumulate is representative of how bad an EKF core has been over time. So now with affinity, we have a way to determine which sensor combination is the best that can be used. For this, the error term (that we accumulate) should be a right indicator of the sensors performance of that core. 

As Tridge pointed out in his affinity PR for EKF2 #11257  - This provides for a more statistically robust way of using multiple sensors when the airframe is known to have multiple "good" sensors. EKF lane changing is used to select the lane with the best sensors.

@tridge please correct me if I am wrong or have missed something !

**Comment by tridge** [comment]
> @tridge please correct me if I am wrong or have missed something !

exactly right, and good explanation. We'll use this in the docs we'll need to write for this feature :-)

**Comment by tridge** [comment]
@harshitsankhla as Paul says, we need to setup some tests to see if we have the tuning right. One of the key properties of the lane switching is to save the vehicle on IMU failures. I sent you a log where the first gyro stops responding to vehicle movement on takeoff. The current master lane switching code wasn't fast enough to save the vehicle from flipping over. 
Can you setup some tests in SITL with a few types of major IMU failures. For example:
 - freezing gyro, like in the log I sent you
 - major IMU failure, such as suddenly accel Z axis changes by 5m/s/s or more (which can happen with aliasing)

what we're looking for is to see if the new lane switching scheme is at least as fast as master for these failures.
We also need to organise some real flight tests to ensure we don't get too many lane swicthes

**Comment by tridge** [comment]
@harshitsankhla any progress on the testing or test suites? See my comment above from last week

**Comment by harshitsankhla** [comment]
> @harshitsankhla any progress on the testing or test suites? See my comment above from last week

here - #14829

**Comment by harshitsankhla** [comment]
> @harshitsankhla any progress on the testing or test suites? See my comment above from last week

here - #14829

**Comment by harshitsankhla** [comment]
moved to #14674
